Add missing references

davidlt opened the pull request:
The followin symbols:
- `HcalDeterministicFit::HcalRegion`
- `HcalDeterministicFit::negthre`
- `HcalDeterministicFit::invGpar`

are used in `void HcalDeterministicFit::apply<T>` methods.

@davidlange6 this should resolve issues on Clang builds. Possible the last bit.

Signed-off-by: David Abdurachmanov David.Abdurachmanov@cern.ch

cmsbuild commented:
A new Pull Request was created by @davidlt  for CMSSW_7_5_X.

Add missing references

It involves the following packages:

RecoLocalCalo/HcalRecAlgos


@cmsbuild, @cvuosalo, @slava77 can you please review it and eventually sign? Thanks.
@argiro this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented on RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc:
```diff
@@ -3,6 +3,10 @@
+const float HcalDeterministicFit::invGpar[3];
+const float HcalDeterministicFit::negthre[2];
+const int HcalDeterministicFit::HcalRegion[2];
+
```
why is landauFrac ( of ``` static constexpr float landauFrac[] { initialization } ``` ) from the header file doesn't have to be in .cc?
All of these are constexpr in .h. Should a text object still be defined in the object file/library?

davidlt replied:
Depends on their usage. C++ allows to define and initialize variables in class/struct, but if it's odr-used in close it must generate the symbol. In that case standard says that compiler diagnostic is not required IIRC, i.e., symbol not generated, linker is not happy. I don't know if constexpr gives an extra play here.

Read C++11 standard 9.4.2/3 and 9.4.2/4. 

Or C++14 (as we are using it now): http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4296.pdf

Sames sections.

davidlt replied:
The most common problem in CMSSW is attempting to take a reference of such thing, then it's odr-used. I have not looked at the code to exactly figure out the location where those data members could be odr-used.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 6b4786fa929f1eb155d35b89f8106f63990fb8bc
 I found an error when building: 
 
 <pre>>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-12-1100/src/RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-12-1100/src/RecoLocalCalo/HcalRecAlgos/src/HcalHFStatusBitFromDigis.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-12-1100/src/RecoLocalCalo/HcalRecAlgos/src/HcalHFStatusBitFromRecHits.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-12-1100/src/RecoLocalCalo/HcalRecAlgos/src/HcalHF_PETalgorithm.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-12-1100/src/RecoLocalCalo/HcalRecAlgos/src/HcalHF_S9S1algorithm.cc 
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-12-1100/src/RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc:6:44: error: 'constexpr' needed for in-class initialization of static data member 'const float HcalDeterministicFit::invGpar [3]' of non-integral type [-fpermissive]
 const float HcalDeterministicFit::invGpar[3];
                                            ^
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-12-1100/src/RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc:7:44: error: 'constexpr' needed for in-class initialization of static data member 'const float HcalDeterministicFit::negthre [2]' of non-integral type [-fpermissive]
 const float HcalDeterministicFit::negthre[2];
                                            ^
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9599/5469/summary.html

slava77 commented:
so, it looks like constexpr is not a const

davidlt commented:
I am working on this now. But for reference `constexpr` changed in C++14, i.e. it does not imply `const` on functions at least.

Commit:
Add missing references

The followin symbols:
- `HcalDeterministicFit::HcalRegion`
- `HcalDeterministicFit::negthre`
- `HcalDeterministicFit::invGpar`

are used in oid HcalDeterministicFit::apply<T>methods.

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>

cmsbuild commented:
Pull request #9599 was updated. @cmsbuild, @cvuosalo, @slava77 can you please check and sign again.

davidlt commented:
Changed `const` to `constexpr` (tested on Clang and GCC). Actually I cannot recall why I put `const` in a first place. I will relaunch the PR testing in Jenkins.

cmsbuild commented:
+1
Tested at: 7480de8eaf8f931a2ed2acbf33b087ff5a2e98d4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9599/10331/summary.html

slava77 commented:
It looks like the comparisons script failed
https://cmssdt.cern.ch/jenkins/job/compare-root-files-short-matrix/8183/console

davidlt commented:
Because there are no baseline results. I will relaunch testing and it should use a different IB by now.

cmsbuild commented:
+1
Tested at: 7480de8eaf8f931a2ed2acbf33b087ff5a2e98d4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9599/5500/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9599/5500/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/9599	7480de8
- code changes are simple and don't affect execution (at least in normal builds)
- jenkins tests passed and comparisons with baseline show no differences as expected

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_5_X IBs  (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

davidlange6 commented:
+1
